server/etcdserver/v3_server.go

```diff
@@ -831,6 +827,59 @@ func (s *EtcdServer) linearizableReadLoop() {
 	}
 }
 
+func (s *EtcdServer) readIndexLoop() {
+	for {
+		requestId := s.reqIDGen.Next()
+		leaderChangedNotifier := s.leaderChanged.Receive()
+		select {
+		case <-leaderChangedNotifier:
+			continue
+		case <-s.readIndexWaitc:
+		case <-s.stopping:
+			return
+		}
+
+		nextnr := newNotifier()
+		s.readMu.Lock()
+		nr := s.readIndexNotifier
+		s.readIndexNotifier = nextnr
+		s.readMu.Unlock()
+
+		confirmedIndex, err := s.requestCurrentIndex(leaderChangedNotifier, requestId)
+		if isStopped(err) {
+			return
+		}
+		if err != nil {
+			nr.notify(err)
+			continue
+		}
+		nr.uint64Val = confirmedIndex
+		// unblock all read index requested at indices before confirmedIndex
+		nr.notify(nil)
+	}
+}
+
+func (s *EtcdServer) GetReadIndex(ctx context.Context) (uint64, error) {
```

Comment:
What's the difference between `GetReadIndex` and `LinearizableReadNotify`?

Reply:
@siyuanfoundation Is there any counter argument of this above comment? Could you please clarify?

Reply:
oh, misread that the new `readIndexLoop` won't wait for the applied index catch up. 

Right after read index is received in `linearizableReadLoop`, `readIndexNotifier` would notify the caller. I think that simply the implementation, wdyt? 

We can reduce the background go routine, avoid `readMu` lock twice and don't need to add index into the notifier struct.